mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/BlockMatrix.scala

```diff
@@ -182,6 +184,38 @@ class BlockMatrix(
     this
   }
 
+  /** Converts to CoordinateMatrix. */
+  def toCoordinateMatrix(): CoordinateMatrix = {
+    val entryRDD = blocks.flatMap { case ((blockRowIndex, blockColIndex), mat) =>
+      val rowStart = blockRowIndex.toLong * rowsPerBlock
+      val colStart = blockColIndex.toLong * colsPerBlock
+      mat match {
+        case dn: DenseMatrix =>
+          val entryValues = new ArrayBuffer[MatrixEntry](mat.numRows * mat.numCols)
```

Comment:
Maybe we can use `mat.foreachActive` directly and start with an empty `ArrayBuffer`. There will be some overhead of allocating new buffers, but it should be small. The code will get simplified.